File: src/main/parsers/duplicate.ts
// Duplicate code detection: sliding window of N normalized lines, hashed.
// Reports occurrences sharing the same hash across files.

import { createHash } from 'node:crypto';
import { DEFAULT_DUPLICATE_LINES } from '../../shared/api';

export interface DupSlice {
  hash: string;
  startLine: number;
  endLine: number;
}

const MIN_SUBSTANTIVE_LINES = 2;

function normalize(line: string): string {
  return line.replace(/\s+/g, ' ').trim();
}

function stripJsxTags(line: string): string {
  return line
    .replace(/<>|<\/>/g, '')
    .replace(/<\/?[A-Za-z][\w.:\-]*(?:\s+[^<>]*)?\/?\s*>/g, '');
}

function isCommentLine(line: string): boolean {
  return /^(?:\/\/|\/\*|\*\/|\*|#)/.test(line);
}

function isImportOrDeclarationLine(line: string): boolean {
  return /^(?:import|export\s+import|use|namespace|require(?:_once)?|include(?:_once)?)\b/i.test(line)
    || /^(?:export\s+)?(?:abstract\s+|final\s+)?(?:class|interface|trait|enum|record|module)\b/i.test(line);
}

function isCallableDeclarationLine(line: string): boolean {
  if (/\bfunction\b/.test(line)) return true;
  if (/^(?:async\s+)?[A-Za-z_$][\w$]*\s*\([^;=]*\)\s*\{$/.test(line) && !/^(?:if|for|while|switch|catch|foreach)\b/.test(line)) {
    return true;
  }
  return /^(?:(?:public|protected|private|internal|static|final|abstract|async|override|virtual|sealed|readonly)\s+)+[A-Za-z_$][\w$<>\[\]?|:&\\]*\s+[A-Za-z_$][\w$]*\s*\([^;=]*\)\s*(?::\s*[A-Za-z_$][\w$<>\[\]?|:&\\]*)?\s*\{?$/.test(line);
}

function isBoundaryLine(line: string): boolean {
  return line === '' || isCommentLine(line) || isImportOrDeclarationLine(line) || isCallableDeclarationLine(line);
}

function isStructuralLine(line: string): boolean {
  if (line === '') return true;
  if (/^[{}()[\];,]+$/.test(line)) return true;
  if (/^return\s*[({[]$/.test(line)) return true;
  const withoutJsx = stripJsxTags(line).replace(/[{}()[\];,]/g, '').trim();
  if (withoutJsx === '') return true;
  return false;
}

function isSubstantiveLine(line: string): boolean {
  return line !== ''
    && !isStructuralLine(line)
    && !isCommentLine(line)
    && !isImportOrDeclarationLine(line)
    && !isCallableDeclarationLine(line);
}

export function findDuplicateSlices(content: string, windowSize = DEFAULT_DUPLICATE_LINES): DupSlice[] {
  const normalizedWindowSize = Math.max(3, Math.floor(windowSize));
  const lines = content.split(/\r\n|\n|\r/).map(normalize);

  const segments: number[][] = [];
  let currentSegment: number[] = [];

  for (let index = 0; index < lines.length; index++) {
    const line = lines[index];

    if (isBoundaryLine(line)) {
      if (currentSegment.length > 0) {
        segments.push(currentSegment);
        currentSegment = [];
      }
      continue;
    }

    if (isSubstantiveLine(line)) currentSegment.push(index);
  }

  if (currentSegment.length > 0) segments.push(currentSegment);

  const out: DupSlice[] = [];

  for (const segment of segments) {
    if (segment.length < normalizedWindowSize) continue;

    for (let i = 0; i + normalizedWindowSize <= segment.length; i++) {
      const windowIndexes = segment.slice(i, i + normalizedWindowSize);
      const startIndex = windowIndexes[0];
      const endIndex = windowIndexes[windowIndexes.length - 1];
      const substantiveSlice = windowIndexes.map(index => lines[index]);
      if (substantiveSlice.length < Math.max(MIN_SUBSTANTIVE_LINES, normalizedWindowSize)) continue;
      const h = createHash('sha1').update(substantiveSlice.join('\n')).digest('hex').slice(0, 16);
      out.push({ hash: h, startLine: startIndex + 1, endLine: endIndex + 1 });
    }
  }

  return out;
}

File: src-tauri/src/parsers/duplicate.rs
use sha1::{Digest, Sha1};

#[derive(Debug, Clone)]
pub struct DupSlice {
    pub hash: String,
    pub start_line: i64,
    pub end_line: i64,
}

fn normalize(line: &str) -> String {
    line.split_whitespace().collect::<Vec<_>>().join(" ").trim().to_string()
}

fn is_comment_line(line: &str) -> bool {
    line.starts_with("//") || line.starts_with("/*") || line.starts_with("*/") || line.starts_with('*') || line.starts_with('#')
}

fn is_import_or_decl(line: &str) -> bool {
    let lower = line.to_lowercase();
    lower.starts_with("import")
        || lower.starts_with("export import")
        || lower.starts_with("use ")
        || lower.starts_with("namespace")
        || lower.starts_with("require")
        || lower.starts_with("include")
        || Regexish::is_type_decl(line)
}

struct Regexish;
impl Regexish {
    fn is_type_decl(line: &str) -> bool {
        once_cell::sync::Lazy::<regex::Regex>::force(&TYPE_DECL).is_match(line)
    }
}

static TYPE_DECL: once_cell::sync::Lazy<regex::Regex> = once_cell::sync::Lazy::new(|| {
    regex::Regex::new(r"^(?:export\s+)?(?:abstract\s+|final\s+)?(?:class|interface|trait|enum|record|module)\b").unwrap()
});

fn is_callable_decl(line: &str) -> bool {
    if line.contains("function") { return true; }
    false
}

fn is_boundary(line: &str) -> bool {
    line.is_empty() || is_comment_line(line) || is_import_or_decl(line) || is_callable_decl(line)
}

fn is_structural(line: &str) -> bool {
    if line.is_empty() { return true; }
    let only = line.chars().all(|c| "{}()[];,".contains(c));
    only
}

fn is_substantive(line: &str) -> bool {
    !line.is_empty() && !is_structural(line) && !is_comment_line(line) && !is_import_or_decl(line) && !is_callable_decl(line)
}

pub fn find_duplicate_slices(content: &str, window_size: i64) -> Vec<DupSlice> {
    let window = window_size.max(3) as usize;
    let lines: Vec<String> = content.lines().map(normalize).collect();
    let mut segments: Vec<Vec<usize>> = Vec::new();
    let mut current: Vec<usize> = Vec::new();
    for (index, line) in lines.iter().enumerate() {
        if is_boundary(line) {
            if !current.is_empty() {
                segments.push(std::mem::take(&mut current));
            }
            continue;
        }
        if is_substantive(line) {
            current.push(index);
        }
    }
    if !current.is_empty() { segments.push(current); }

    let mut out = Vec::new();
    for segment in segments {
        if segment.len() < window { continue; }
        for i in 0..=(segment.len() - window) {
            let window_indexes = &segment[i..i + window];
            let start_index = window_indexes[0];
            let end_index = *window_indexes.last().unwrap();
            let joined = window_indexes.iter().map(|&idx| lines[idx].as_str()).collect::<Vec<_>>().join("\n");
            let mut hasher = Sha1::new();
            hasher.update(joined.as_bytes());
            let digest = hasher.finalize();
            let hex = hex::encode(&digest[..8]); // 16 hex chars
            out.push(DupSlice {
                hash: hex,
                start_line: (start_index + 1) as i64,
                end_line: (end_index + 1) as i64,
            });
        }
    }
    out
}
